sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp

```diff
@@ -509,20 +503,28 @@ cdef class EuclideanArgKmin{{name_suffix}}(ArgKmin{{name_suffix}}):
         # which keep tracks of the argkmin.
         for i in range(n_X):
             for j in range(n_Y):
-                heap_push(
-                    heaps_r_distances + i * self.k,
-                    heaps_indices + i * self.k,
-                    self.k,
-                    # Using the squared euclidean distance as the rank-preserving distance:
-                    #
-                    #             ||X_c_i||² - 2 X_c_i.Y_c_j^T + ||Y_c_j||²
+                sqeuclidean_dist_i_j = (
+                    self.X_norm_squared[i + X_start] +
+                    dist_middle_terms[i * n_Y + j] +
+                    self.Y_norm_squared[j + Y_start]
+                )
+
+                sqeuclidean_dist_i_j = (
+                    sqeuclidean_dist_i_j
+                    # Catastrophic cancellation causing -0. and NaN can be present
+                    # in some situations, e.g. when computing d(x_i, y_i) when X is Y.
                     #
-                    (
-                        self.X_norm_squared[i + X_start] +
-                        dist_middle_terms[i * n_Y + j] +
-                        self.Y_norm_squared[j + Y_start]
-                    ),
-                    j + Y_start,
+                    #  Guard against -0.    Guard against NaN
+                    #     v                       v
+                    if 0. < sqeuclidean_dist_i_j == sqeuclidean_dist_i_j else 0.
```

Comment:
It is not easy to see that `sqeuclidean_dist_i_j` becomes 0 with this ternary operator. I would still go with:

```python
if sqeuclidean_dist_i_j < 0 or sqeuclidean_dist_i_j != sqeuclidean_dist_i_j:
    sqeuclidean_dist_i_j = 0.
```

Does performing this check influence performance?

Reply:
I do not think having this more explicit syntax worsen performance as I think the ternary operator is more of a syntactic sugar.